### ISSUE TITLE
Create a design to integrate gRPC to the ABP Framework & all modules

### ISSUE THREAD

**Issue by hikalkan** [abpframework/abp#2882]
- [ ] Investigate "how ABP framework can help developers on gRPC API creation".
- [ ] Implement gRPC endpoints for all the modules.
- [ ] Add a starting point for the startup templates.

**Comment by maliming** [comment]
https://github.com/Cysharp/MagicOnion Looks like a good class library.

**Comment by hikalkan** [comment]
Thanks @maliming it seems a very good library. However it seems a real-time communication system. With this issue, we only want a standard gRPC implementation.

**Comment by mehmetuken** [comment]
> https://github.com/Cysharp/MagicOnion Looks like a good class library.

This library using without .proto files and only use .net apps beetwen. See this issue https://github.com/Cysharp/MagicOnion/issues/159#issuecomment-502506139
I think create shared .proto files project each module. It's helpfull any language implement and communicate each other.

**Comment by geffzhang** [comment]
https://devblogs.microsoft.com/premier-developer/sharing-grpc-protobuf-contracts-using-a-rest-endpoint/

**Comment by geffzhang** [comment]
https://github.com/thangchung/GrpcJsonTranscoder

**Comment by hikalkan** [comment]
Interesting:

![image](https://user-images.githubusercontent.com/1210527/82615225-97c61500-9bd2-11ea-8050-76134ac3474a.png)

https://docs.microsoft.com/en-us/aspnet/core/grpc/basics

**Comment by hikalkan** [comment]
No meaning to implement it if it is not supported by Azure and IIS. We are planning to implement Blazor UI to server communication via gRPC but this is a huge restriction.

**Comment by hikalkan** [comment]
We are deferring this issue.

We were planning to use the gRPC endpoints with the Blazor UI, so wanted to work on this first. However, I see that gRPC with AspNet Core & Blazor is not so production ready. Here, the reasons:

Warning at https://docs.microsoft.com/en-us/aspnet/core/grpc/basics

![image](https://user-images.githubusercontent.com/1210527/82734712-64e06600-9d25-11ea-8848-e64aba4171f5.png)

Warning/Information at https://docs.microsoft.com/en-us/aspnet/core/grpc/browser

![image](https://user-images.githubusercontent.com/1210527/82734730-85102500-9d25-11ea-8449-9cc48cb9c7d3.png)

And some others...

![image](https://user-images.githubusercontent.com/1210527/82734745-a53fe400-9d25-11ea-8688-c9819c01e14e.png)

So, instead of using gRPC endpoints, we will try to use the existing REST endpoints with the Blazor UI. This will speed up our development for the Blazor UI.

Anyone wants to use gRPC in their applications, want to create gRPC endpoints for their microservices, they can still use it. We've prepared a sample application: https://github.com/abpframework/abp-samples/tree/master/GrpcDemo

**Comment by geffzhang** [comment]
gRPC-Web for .NET now available https://devblogs.microsoft.com/aspnet/grpc-web-for-net-now-available/

**Comment by hikalkan** [comment]
@geffzhang that's great! As I can see, it solves most of the problems. We will check it deeply after v3.0.

**Comment by expcat** [comment]
https://github.com/protobuf-net/protobuf-net.Grpc
It is a code-frist base on grpc-dotnet

**Comment by kegumx** [comment]
+1

**Comment by rafaelgfirmino** [comment]
+1

**Comment by charleszh-dev** [comment]
Is there any news?😀

**Comment by barnybadzoo** [comment]
Microsoft switched from ocelot to envoy because of the grpc feature.. see here: https://docs.microsoft.com/de-de/dotnet/architecture/microservices/multi-container-microservice-net-applications/implement-api-gateways-with-ocelot
Any further plans also to switch to envoy ? or what is on roadmap for grpc ?

**Comment by leonkosak** [comment]
@hikalkan any updates on this topic?

**Comment by manishkgupta02** [comment]
+1

**Comment by fanslead** [comment]
any updates on this topic?

**Comment by gterdem** [comment]
There is already a [gRPC integration demo](https://github.com/abpframework/abp-samples/tree/master/GrpcDemo) that you can check. However if you want to use gRPC instead of http within the framework; that'll take some more time. 

We've implemented [Static Client Proxy generation](https://github.com/abpframework/abp/issues/9864) so that microservice gateways are no longer referencing _Http.Api_ layers of other microservices to locate endpoints. This will allow us using any other gateway (like envoy etc). With this feature enabled, we've also plans for sample microservices communicating with each other using gRPC so that we can check if we need to make improvements in framework for it.

**TLDR:** It is in our backlog, we are moving slowly but steadily :)

**Comment by RainingNight** [comment]
> https://github.com/protobuf-net/protobuf-net.Grpc It is a code-frist base on grpc-dotnet

Looks better.

**Comment by hikalkan** [comment]
Hi all,

I've created two articles for ABP & gRPC:

* [Using gRPC with the ABP Framework](https://community.abp.io/posts/using-grpc-with-the-abp-framework-2dgaxzw3)
* [Consuming gRPC Services from Blazor WebAssembly Application Using gRPC-Web](https://community.abp.io/posts/consuming-grpc-services-from-blazor-webassembly-application-using-grpcweb-dqjry3rv)

I see that using gRPC in an ABP-based solution is straightforward. Currently, we don't think to implement any specific integration package. However, I've taken a lot of notes while investigating .NET's gRPC integration and preparing these articles. In the future, we will probably make some built-in integrations to make it even easier. Some of my notes:

* We can provide packages to automatically register services in the server & client side.
* We can also reduce the boilerplate code while creating channels and consuming services.
* We may provide some services to create and configure channels in a central place.
* We can provide client-side proxy generation with also Dapr integration.
* We may somehow automatically add tenant id, current culture, correlation id, authorization token... to gRPC requests.

However, all these are additional ABP benefits and currently there is no difference of using gRPC in ABP by comparing in with a .NET solution without ABP.
